rust/agents/relayer/src/msg/metadata/aggregation.rs

```diff
@@ -0,0 +1,238 @@
+            // Overwrite the 0-initialized buffer
+            buffer.splice(
+                encoded_range_start..(encoded_range_start + METADATA_RANGE_SIZE * 2),
+                [encode_byte_index(range_start), encode_byte_index(range_end)].concat(),
+            );
+        }
+        buffer
+    }
+
+    fn n_cheapest_metas(
+        mut metas_and_gas: Vec<(SubIsmMetadata, U256)>,
+        n: usize,
+    ) -> Vec<SubIsmMetadata> {
+        metas_and_gas.sort_by(|(_, gas_1), (_, gas_2)| gas_1.cmp(gas_2));
+        let mut cheapest: Vec<_> = metas_and_gas[..n].into();
+        cheapest.sort_by(|(meta_1, _), (meta_2, _)| meta_1.index.cmp(&meta_2.index));
+        cheapest.iter().cloned().map(|(meta, _)| meta).collect()
+    }
+
+    async fn cheapest_valid_metas(
+        sub_isms: Vec<(SubIsmMetadata, Box<dyn InterchainSecurityModule>)>,
+        message: &HyperlaneMessage,
+        threshold: usize,
+    ) -> eyre::Result<Option<Vec<SubIsmMetadata>>> {
+        let metas_and_gas: Vec<_> =
+            join_all(sub_isms.into_iter().map(|(sub_ism_meta, ism)| async move {
+                let gas = ism.dry_run_verify(message, &sub_ism_meta.metadata).await;
+                (sub_ism_meta, gas)
+            }))
+            .await
+            .into_iter()
+            .filter_map(|(sub_ism_meta, gast_result)| {
+                gast_result.ok().flatten().map(|gc| (sub_ism_meta, gc))
+            })
+            .collect();
+
+        let metas_and_gas_count = metas_and_gas.len();
+        if metas_and_gas_count < threshold {
+            info!("Could not fetch all metadata: Found {metas_and_gas_count} of the {threshold} required ISM metadata pieces");
+            return Ok(None);
+        }
+        Ok(Some(Self::n_cheapest_metas(metas_and_gas, threshold)))
+    }
+}
+
+impl Deref for AggregationIsmMetadataBuilder {
+    type Target = BaseMetadataBuilder;
+
+    fn deref(&self) -> &Self::Target {
+        &self.base
+    }
+}
+
+#[async_trait]
+impl MetadataBuilder for AggregationIsmMetadataBuilder {
+    #[instrument(err, skip(self))]
+    async fn build(
+        &self,
+        ism_address: H256,
+        message: &HyperlaneMessage,
+    ) -> eyre::Result<Option<Vec<u8>>> {
+        const CTX: &str = "When fetching RoutingIsm metadata";
+        let ism = self.build_aggregation_ism(ism_address).await.context(CTX)?;
+        let (modules, threshold) = ism.modules_and_threshold(message).await.context(CTX)?;
+        let threshold = threshold as usize;
+
+        let sub_isms: Vec<_> = join_all(modules.iter().map(|ism_address| {
+            try_join(
+                self.base.build(*ism_address, message),
+                self.base.build_ism(*ism_address),
+            )
+        }))
+        .await
+        .into_iter()
+        .enumerate()
+        .filter_map(|(i, r)| match r {
+            Ok((Some(meta), ism)) => Some((SubIsmMetadata::new(i, meta), ism)),
+            _ => None,
+        })
```

Comment:
this is hard to read
why do we need to build the metadata and ISM here?

Reply:
The `try_join` above saves some computation by stopping early if either of the futures returns an error. You're right that it's too complicated for what it does though, I'll simplify